controller/proxy-injector/webhook.go

```diff
@@ -148,3 +159,36 @@ func (w *Webhook) inject(request *admissionv1beta1.AdmissionRequest) (*admission
 
 	return admissionResponse, nil
 }
+
+func (w *Webhook) parentRefLabel(ns string, ref metav1.OwnerReference) (string, string, error) {
```

Comment:
We already have a `GetOwnerKindAndName` func for finding pod owners in controller/k8s/api.go, and my preference would be to use that here.

https://github.com/linkerd/linkerd2/blob/4eb89bb8c22fa161be801038c090c3796d60f025/controller/k8s/api.go#L328-L347

In order to do that, you'll need to wire through a `k8s.API` object from `controller/cmd/proxy-injector/main.go` to this file. My suggestion would be to create it in main.go as follows:

```
	k8sAPI, err := k8s.InitializeAPI(*kubeconfig, k8s.NS, k8s.RS)
```

So that it automatically watches namespaces and replicasets. That will also allow us to replace our uncached namespace reads in this file with cached reads (e.g. `k8sAPI.NS().Lister().Get(request.Namespace)`).

Reply:
@klingerf I did it like that because the admission request gets me the `ownerReferences` JSON outside of the pod spec. But I can simply manufacture a proper `corev1.Pod` settings by setting those refs manually, and pass it to `GetOwnerKindAndName` (whose caching properties I was unaware of, so that'll be an important gain). On the other hand, `GetOwnerKindAndName` doesn't consider Replication Controllers; should I just add support for that in that function?

Reply:
Good questions. We don't call `GetOwnerKindAndName` in too many places, so I'd be OK with changing its method signature to make it more accommodating for your use case. Am pretty sure we could get by with passing `pod.ObjectMeta` to that function, instead of the full `Pod` object.

`GetOwnerKindAndName` already supports ReplicationControllers. All types of workloads can be represented in the `OwnerReference` `Kind` field. If the `Kind` is set to "ReplicationController", we'll return the `Kind` and `Name` fields without needing to look up the full ReplicationController object. Only if the `Kind` is set to "ReplicaSet" do we need to do an additional lookup to get the full ReplicaSet object, so that we can get its owner references, which will give us the Deployment. For an example of how it works, see the `GetOwnerKindAndName` tests, here:

https://github.com/linkerd/linkerd2/blob/master/controller/k8s/api_test.go#L588

Reply:
Gotcha. I was under the impression that `ReplicationControllers` could themselves have an owner reference, but that's not the case :grimacing: